Implement IntakeSubsystem and IntakeCommand

FireShredder24 opened the pull request:
Implements a Subsystem that controls the 2 AM-RedLine motors that spin the intake rollers and the two pneumatic pistons that actuate the intake, as well as a Command that binds said Subsystem to the controller.

FireShredder24 committed:
Add IntakeSubsystem

With runIntake() and stopIntake() methods for ease of use.

FireShredder24 committed:
Add IntakeCommand

Takes an instance of RobotContainer as a constructor to avoid static references to members of RobotContainer.

FireShredder24 committed:
Add IntakeCommand to RobotContainer

FireShredder24 committed:
Implement Intake Actuation

Implements the 2 DoubleSolenoids that actuate the Power Cell intake between its starting and ready positions.
Updates the controller class and interface to include get()s for those functions, but without a button mapping.
Updates constants to include PCM channels.

FireShredder24 committed:
Change intake lowering to a toggle

Changes the control of the intake position from a 2-button approach to a single-button toggle.

terbos90803 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -32,7 +32,13 @@
+  public static final int INTAKE_VICTORSP_PWM_CHAN_LF = 3;
+  public static final int INTAKE_VICTORSP_PWM_CHAN_RT = 4;
+  // Pneumatic solenoid addresses
+  public static final int INTAKE_PISTON_PCM_CHAN_LF_F = 1;
```
Consider renaming with higher level sense of direction.
Instead of F and R for forward and reverse, perhaps choose DEPLOY and STOW.

terbos90803 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,71 @@
+ * An example command that uses an example subsystem.
+ */
+public class IntakeCommand extends CommandBase {
+  @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
```
Don't suppress warnings at this time.  You need to be reminded that the code is not complete.

jordan-powers replied:
We don't even use the PMD analyzer. I'm guessing this is left over from an example somewhere (I'm pretty sure wpilib uses PMD).

FireShredder24 replied:
Yeah, that was from the ExampleCommand.

terbos90803 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,71 @@
+public class IntakeCommand extends CommandBase {
+  @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+  private final IntakeSubsystem m_subsystem;
+  private RobotContainer m_container;
```
Should be final.

terbos90803 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,71 @@
+  private boolean isLowered = false;
+
+  /**
+   * Creates a new ExampleCommand.
```
Update comment

terbos90803 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,71 @@
+  // Called every time the scheduler runs while the command is scheduled.
+  @Override
+  public void execute() {
+    if (m_container.mainController.getIntakePowerCells()) {
```
I know that the example code from last year did this kind of things where a command would keep running and then poll the joystick buttons.  While this does work, the intended usage of the Button commands is to trigger on a button press, update the subsystem state, and then end.
It's more reasonable for the Drive to run continuously since it's reading the analog inputs, but the buttons can be broken out separately.
We should talk about this.

terbos90803 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,71 @@
+  private final IntakeSubsystem m_subsystem;
+  private RobotContainer m_container;
+
+  private boolean isLowered = false;
```
You should not hold state in the Command.  State should be held in the subsystem.

terbos90803 commented on src/main/java/frc/robot/controllers/DriveController.java:
```diff
@@ -36,6 +36,13 @@
+   * @return Whether the pistons should fire that raise the intake back to its
+   *         starting position.
+   */
+  abstract public boolean getToggleIntake();
```
The abstract keyword is not needed for interfaces as it's implied.

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -7,14 +7,42 @@
+  }
+
+  public void runIntake() {
+    intakeSPrt.set(0.3);
```
1) No magic numbers.  Use named constants.
2) Should move motor set calls into periodic method.
3) Better API is: public void runIntake(bool run);

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -7,14 +7,42 @@
+  }
+
+  public void raiseIntake() {
+    intakePistonL.set(DoubleSolenoid.Value.kForward);
```
Should replace reference to kForward etc with class constants to abstract Deploy vs Stow.

FireShredder24 committed:
Fix interface implementation error

FireShredder24 committed:
Apply PR suggestions

Move state to Subsystem from the Command
Change the intake roller control to reflect the fact that motor control is per-packet, not set()-and-forget.
Rename piston constants to reflect a higher sense of direction.
Do the same for the solenoid values in the Subsystem.

FireShredder24 committed:
Bind Commands to JoystickButtons

Switch to the intended use of JoystickButtons in WPILib, where a Command is auto-scheduled based on a specified state of a button.
Remove unnecessary "abstract" prefixes from DriveController.
Add toggle-reversal functionality to the intake rollers.
Add toggle functionality to the intake pistons.

terbos90803 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -36,9 +36,9 @@
+  public static final int INTAKE_PISTON_PCM_CHAN_LF_DEPLOY = 1;
+  public static final int INTAKE_PISTON_PCM_CHAN_LF_STOW = 2;
+  public static final int INTAKE_PISTON_PCM_CHAN_RT_DEPLOY = 3;
+  public static final int INTAKE_PISTON_PCM_CHAN_RT_STOW = 4;
```
PCM channels are numbered 0-7.

terbos90803 commented on src/main/java/frc/robot/commands/intake/IntakePistonToggle.java:
```diff
@@ -5,33 +5,25 @@
  */
-public class IntakeCommand extends CommandBase {
-  @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+public class IntakePistonToggle extends CommandBase {
```
Blanket comment to all Commands:  The name of the class should have Command or at least Cmd in it.  This promotes readability of the calling code.  Whichever you choose, be consistent across all instances.

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -16,37 +16,59 @@
   @Override
   public void periodic() {
     // This method will be called once per scheduler run
+    if (runIntake && !reverseIntake) {
```
Calculate motor setpoint first, then always set the motor value, even if it's zero.
If you don't set the motor, you get a bunch of watchdog warnings.

FireShredder24 committed:
Apply PR Suggestions

Rename the intake commands to be more readable and clear
Change PCM mappings to be more accurate
Prevent watchdog errors by setting the intake roller motors to 0 if runIntake is false.
Break out the intake roller setpoint into a higher-level constant.

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -24,6 +24,7 @@
   public boolean isLowered = false;
   private boolean runIntake = false;
   private boolean reverseIntake = false;
+  private final double rollerSetpoint = 0.3;
```
Should be a MoPref so we can tune it from the ShuffleBoard.

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -64,11 +65,14 @@ public void lowerIntake() {
+      intakeSPrt.set(-rollerSetpoint);
+    } else {
+      intakeSPlf.set(0);
+      intakeSPrt.set(0);
```
Good coding practice is to only set the motor values in one place.  Do all the math/selection first, then set the values once.

Example:
```
double setpoint = runIntake ? rollerSetpoint : 0;
if (reverseIntake)
  setpoint = -setpoint;

intakeSPlf.set(setpoint);
intakeSPrt.set(setpoint);
```

FireShredder24 committed:
Break out RollerSetpoint into MoPrefs

FireShredder24 committed:
Initialize MoPrefs in RobotContainer

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -25,13 +35,24 @@
+  private final ControllerBase mainController = new ControllerBase(xbox, f310);
 
-  public final ControllerBase mainController = new ControllerBase();
+  private final JoystickButton intakeRollerFwdButton = new JoystickButton(f310, 4/* LeftBumper */);
+  private final JoystickButton intakeRollerFwdRevToggle = new JoystickButton(f310, 0/* X */);
+  private final JoystickButton intakePistonToggle = new JoystickButton(f310, 2/* B */);
```
There should be enums for the named buttons instead of numbers.

terbos90803 commented on src/main/java/frc/robot/commands/intake/IntakePistonCmd.java:
```diff
@@ -0,0 +1,53 @@
+/*----------------------------------------------------------------------------*/
```
Each of these command files is 99% boilerplate and 1 line of useful code.  These should all be converted to lambdas in configureButtonBindings.

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -7,18 +7,72 @@
-  /**
-   * Creates a new ExampleSubsystem.
-   */
+  private final VictorSP intakeSPrt = new VictorSP(Constants.INTAKE_VICTORSP_PWM_CHAN_LF);
```
Robot System says there is only one PWM mapped for both motors, so only need one Victor class.

FireShredder24 committed:
Use inline InstantCommands to control intake

Removes the extraneous Commands that were used for instant single-method execution and replaces them with inline InstantCommands in RobotContainer.
Remaps the PWM channels of the intake roller controllers to match the system connections chart, including condensation into a single VictorSP.

terbos90803 reviewed (approved):
Ok.  I think that's close enough to go.

terbos90803 committed:
Merge branch 'master' into IntakeSubsystem
